Add Plugin Tag Cloud

tkelley-eg opened the pull request:
- Created new PluginLabelTagCloud component.
- Added graphql query to get labels and transform to structure required by new TagCloud component.
- Added component to index and need styling help.

Related to issue #137  

Summary of this pull request: 

I plopped this on the main page, but could use a little styling advice.  Is the vertical location good?  Should we create some margins for whitespace on the sides?

For the search page, should it be placed at the bottom on the left criteria area?

![image](https://user-images.githubusercontent.com/3812023/96367996-16da5d80-111f-11eb-9cbb-b56b99c00113.png)

tkelley-eg committed:
Created new PluginLabelTagCloud component.
Added graphql query to get labels and transform
to structure required by new TagCloud component.
Added component to index and need styling help.

tkelley-eg committed:
Removed debugger statement.

tkelley-eg commented on src/components/PluginLabelTagCloud.jsx:
```diff
@@ -0,0 +1,50 @@
+    }, {});
+
+    const tags = new Array();
+    Object.entries(allLabels).forEach(([value, count]) => {
```
It wasn't obvious to me how to create this structure while doing the `reduce` ops above....

zbynek replied:
Maybe it can be simplified by converting the inner `reduce` to `forEach` (you don't really need `prevLabel` for anything)

halkeye replied:
i think this is fine, but you could map it

```
const tags = Object.entries(allLabels).map(([value, count]) => { return { value, count } });
```

halkeye commented:
> I plopped this on the main page, but could use a little styling advice. Is the vertical location good? Should we create some margins for whitespace on the sides?

Do you have a screenshot for this? I'm so bad with whitespace but I'd like to see it in action (my dev environment is offline right now so i can't pull it)

> For the search page, should it be placed at the bottom on the left criteria area?

I'm not really sure it should go on the search page at all, just speaking out loud here, what about the main page, and a dedicated /ui/tagcloud page?

halkeye commented on src/components/PluginLabelTagCloud.jsx:
```diff
@@ -0,0 +1,50 @@
+            minSize={12}
+            maxSize={35}
+            tags={tags}
+            onClick={tag => document.location.href=`/ui/search/?labels=${tag.value}`}
```
Instead of changing the location href, which will work, you should use the gatsby helper - https://www.gatsbyjs.com/docs/gatsby-link/#how-to-use-the-navigate-helper-function

This will prevent the entire page load if data is cached and other things.

halkeye commented on src/components/PluginLabelTagCloud.jsx:
```diff
@@ -0,0 +1,50 @@
+    `).labels.nodes.reduce((prev, nodes) => {
+
+        nodes.labels.reduce((prevLabel, obj) => {
+            if (!prev[obj]) {
```
just fyi, i'm not sure its any clearer to read, you can do it in one line

```
prev[obj] = (prev[obj] || 0) + 1
```

zbynek reviewed (commented):
Thanks for the PR. Maybe @oleg-nenashev can comment on the desired position of this tag cloud. Depending on the position we should adjust the colors used by the cloud -- currently it's too light to be readable on white background.

zbynek commented:
@halkeye  there's the screenshot:

![image](https://user-images.githubusercontent.com/1105305/96375187-ba575c80-1177-11eb-9ec1-98fc8aae7e0f.png)

For margins it would be better to apply some existing classname from Bootstrap; I think `container` should do the trick.

halkeye commented:
> @halkeye there's the screenshot:

oh thats slick. Some of the colors are hard on the eyes on the white bg, and I think it needs context, but its a nice addition

tkelley-eg committed:
Optimized some structure building statements.
Added gastby navigate for links.
Added cursor on hover to help id clickable links
Added blocked list of names to exclude from cloud.

tkelley-eg commented:
Is there anything I might do to help bring this one across the finish line?  Thanks!

zbynek commented:
Given @daniel-beck 's comment I'm not sure what's the best way forward here -- there are many more labels that should be filtered (all labels that are used for a single plugin, labels like `api-plugin`, probably also `builwrapper`, `post-build`, `ui`, `slaves`, `agent`...), in some cases synonyms should be merged (stash = bitbucket = bitbucket-server). Maybe what we need to achieve @oleg-nenashev 's original goal (highlighting useful plugins from different areas) is not a word cloud at all.

I think it's good that thanks to this PR we have an idea about the used labels and their frequencies, but maybe the UX sig could chime in before you spend time perfecting a feature that's gonna be removed.

oleg-nenashev commented:
Sorry. It was not in my review list and I have missed a few pings. Will try to review it over the weekend and come up with suggestions. 

I envisioned a relatively small tag cloud pane under the existing quick filters, something like that:

![image](https://user-images.githubusercontent.com/3000480/97696805-623d2600-1aa6-11eb-874f-07cf3db13aac.png)
.

Maybe we could also put it under the search to utilize the currently unused space on desktop resolutions

![image](https://user-images.githubusercontent.com/3000480/97696986-a9c3b200-1aa6-11eb-9d20-bdc8252f31b1.png)

fqueiruga commented on src/components/PluginLabelTagCloud.jsx:
```diff
@@ -0,0 +1,51 @@
+        return prev;
+    }, {});
+
+    const tags = Object.entries(allLabels).map(([value, count]) => {return {value, count};});
```
`Object.entries` is not supported on IE https://caniuse.com/object-entries.  I think a polyfill is needed or use `for .. in` instead..

halkeye replied:
everything in gastby will go through babel so shouldn't need to worry.

That being said, one of the dependancies breaks on ie10 because it uses Map and doesn't seem to be babeled - https://app.saucelabs.com/tests/b9d8dc52f6354636b481ca0f73f85a22
Looks like also breaks in ie11, but different place, somewhere in react - https://app.saucelabs.com/tests/64266ab032ab42f181a9b1131e7db982

That being said, the site is designed to export regular html with javascript applied on top after its finished loading, so the older browsers won't get the full experience, but they will get the info. I wouldn't worry about object.entries

halkeye commented:
For those who want to see it live:
Unique Deploy URL: https://5f9c4ccad4709a0c81d4fded--pensive-galileo-3fa514.netlify.app

zbynek committed:
Merge branch 'master' of https://github.com/jenkins-infra/plugin-site into 137-label-tag-cloud

zbynek commented:
This has been inactive for a long time, if someone strogly feels this is worth implementing, feel free to reopen.
